src/condor_starter.V6.1/starter.cpp

```diff
@@ -2077,10 +2081,309 @@ Starter::jobEnvironmentCannotReady(int status, const struct UnreadyReason & urea
+					diagnosticResultAd.InsertAttr( "ExitStatus", status );
+				}
+			}
+		}
+	}
+
+
+	//
+	// Reply.
+	//
+	SEND_REPLY_AND_CONTINUE_CONVERSATION;
+}
+
+
+condor::cr::void_coroutine
+retrySetupJobEnvironment(JobInfoCommunicator * jic) {
+	// This is a hack, but setting up `co_yield 0;` to do the right thing
+	// would be a lot of work and lack flexibility, plus be a little bit
+	// of an abuse of how system is supposed to work.
+	condor::dc::AwaitableDeadlineSocket ads;
+	ads.deadline( nullptr, 0 );
+
+	// This seems to work, which is moderately terrifying,
+	// because it's never previously been called twice.
+	jic->setupJobEnvironment();
+
+	co_return;
+}
+
+
+void
+Starter::requestGuidanceJobEnvironmentReady( int /* timerID */ ) {
+	ClassAd request;
+	ClassAd guidance;
+	request.InsertAttr(ATTR_REQUEST_TYPE, RTYPE_JOB_ENVIRONMENT);
+
+	GuidanceResult rv = GuidanceResult::Invalid;
+	if( jic->genericRequestGuidance( request, rv, guidance ) ) {
+		if( rv == GuidanceResult::Command ) {
+			auto lambda = [=, this] (int i) { this->requestGuidanceJobEnvironmentReady(i); };
+			if( handleJobEnvironmentCommand( guidance, lambda ) ) { return; }
+		} else {
+			dprintf( D_ALWAYS, "Problem requesting guidance from AP (%d); carrying on.\n", static_cast<int>(rv) );
+		}
+	}
+
+	// Carry on.
+	this->jobWaitUntilExecuteTime();
+}
+
+
+bool
+Starter::handleJobEnvironmentCommand(
+  const ClassAd & guidance,
+  std::function<void(int)> continue_conversation
+) {
+	std::string command;
+	if(! guidance.LookupString( ATTR_COMMAND, command )) {
+		dprintf( D_ALWAYS, "Received guidance but didn't understand it; carrying on.\n" );
+		dPrintAd( D_ALWAYS, guidance );
+
+		return false;
+	} else {
+		dprintf( D_ALWAYS, "Received the following guidance: '%s'\n", command.c_str() );
+		if( command == COMMAND_START_JOB ) {
+			dprintf( D_ALWAYS, "Starting job as guided...\n" );
+			// This schedules a zero-second timer.
+			this->jobWaitUntilExecuteTime();
+			return true;
+		} else if( command == COMMAND_RETRY_TRANSFER ) {
+			dprintf( D_ALWAYS, "Retrying transfer as guided...\n" );
+			// This schedules a zero-second timer.
+			retrySetupJobEnvironment(jic);
+			return true;
+		} else if( command == COMMAND_RUN_DIAGNOSTIC ) {
+			std::string diagnostic;
+			if(! guidance.LookupString(ATTR_DIAGNOSTIC, diagnostic)) {
+				dprintf( D_ALWAYS, "Received guidance 'RunDiagnostic', but could not find a diagnostic to run; carrying on, instead.\n" );
+
+				return false;
+			} else {
+				dprintf( D_ALWAYS, "Running diagnostic '%s' as guided...\n", diagnostic.c_str() );
+
+				run_diagnostic_reply_and_request_additional_guidance(
+					diagnostic, jic, continue_conversation
+				);
+
+				return true;
+			}
+		} else if( command == COMMAND_ABORT ) {
+			dprintf( D_ALWAYS, "Aborting job as guided...\n" );
+			this->deferral_tid = daemonCore->Register_Timer(0,
+				(TimerHandlercpp)&Starter::SkipJobs,
+				"SkipJobs",
+				this
```

Comment:
Another call site I'd like to see us use the `std::function` overload of `Register_Timer`.

It isn't changed in this PR, but if we could change the Awaitable... coroutines to use the `std::function` overloads, and add a `std::function` overload to `Register_Socket`, then those coroutines would not need to inherit from `public Service`

Reply:
Fixed.  I've written the excellent suggestion up as HTCONDOR-2813.